### ISSUE TITLE
More futures - Some useful concepts here!

### ISSUE THREAD

**Issue by apapacy** [wellyshen/react-cool-starter#77]
This is a great project. But I would like some more buns.

1) deferred actions need to be simplified, for example, as discussed with Dan Abramov, and was implemented https://github.com/erikras/react-redux-universal-hot-example/blob/master/src/redux/middleware/clientMiddleware.js

2) I'd like to upload data for server rendering without additional requests, for example using this technique
not
`
data = await request.get('/test/rest');
`
but
`
app.handle({...request, url: '/test/rest', method: 'get'}, response, next));
router.get('/test/rest', (req,res)=>{
  if (__INNER_REQUEST__) {
    return obj;
  } else {
    res.json(obj);
  }
})
`
3) write the repository to the local storage and restore it when the page is loaded for a friendlier operation of the application in case of pressing the f5 / overload key (after all, not all the data is immediately synchronized with the server)

4) More universal data preloading with server rendering like https://github.com/erikras/react-redux-universal-hot-example/blob/master/src/server.js#L92

**Comment by wellyshen** [comment]
@apapacy

Would you mind to help me to improve it :)?
BTW, I think [redux-thunk](https://github.com/gaearon/redux-thunk) has handled the 1.

**Comment by apapacy** [comment]
Yes, of course. (1) in progress in my fork branch dev-promised: https://github.com/apapacy/react-cool-starter/blob/dev-promised/src/redux/promisedMiddleware.js
More interesting is (4). But this is difficult because of the synchronous nature of the function renderToString. And all more or less working solutions have become irrelevant after new version react/redux/router were released. 
Now I'm try to use https://github.com/markdalgleish/redial for universal data preloading. This library seems as agnostic to use with any release of react/redux/router/

**Comment by wellyshen** [comment]
@apapacy What's the benefit of (1)?

**Comment by apapacy** [comment]
This code with "promisedMiddleware" seems more simple:


```
export const fetchUsers = URL => {
    promise: client => client.get(URL),
    events: [USERS_REQUESTING, USERS_SUCCESS, USERS_FAILURE],
};
```


then with  redux-thunk:


```
export const fetchUsers = (axios: any, URL: string = API_URL): ThunkAction =>
  (dispatch: Dispatch) => {
    dispatch({ type: USERS_REQUESTING });

    return axios.get(URL)
      .then((res) => {
        dispatch({ type: USERS_SUCCESS, data: res.data });
      })
      .catch((err) => {
        dispatch({ type: USERS_FAILURE, err });
      });
};
```

It seems to me that this will work for 90% of the charges.
In some cases, logic can be more complex and in this case you can use redux-thunk

**Comment by wellyshen** [comment]
@apapacy Does the "promisedMiddleware" is necessary for every developer and does it easy to be tested. I wish to make this starter simple and easy to be understood. If it's please help me improve it :)

**Comment by apapacy** [comment]
Asynchronous actions (interaction with the server) for applications with which I worked were in the majority. Working with promises that are hidden from the user greatly simplified my code.

I'll try to make examples and tests in the form of a pool request in your repository.

**Comment by wellyshen** [comment]
@apapacy I'm really appreciate what you did for this improvement. After I merged it, I found the `promiseMiddle` causes [flow](https://flow.org/en/docs/frameworks/react/) hard to track the type of the promise actions. Although the `promiseMiddleware` is convenient for developers. But the type checking by flow is the necessary feature of this starter. Therefore I keep the `promiseMiddle` as the optional choice for developers.
